File: c/math/collatz.c
/* Print the start below 1000 with the longest Collatz sequence, tracked in a running maximum. */
/* Build: cc -o collatz collatz.c */

#include <stdio.h>

#define LIMIT 1000

static int chain_length(long start)
{
    int length = 1;

    while (start != 1) {
        start = start % 2 == 0 ? start / 2 : start * 3 + 1;
        length++;
    }
    return length;
}

int main(void)
{
    int longest = 1;
    int best = 1;
    int start;

    for (start = 1; start < LIMIT; start++) {
        int length = chain_length(start);

        if (length > best) {
            longest = start;
            best = length;
        }
    }
    printf("%d %d\n", longest, best);
    return 0;
}

File: zig/math/collatz.zig
// Print the start below 1000 with the longest Collatz sequence, tracked in a running maximum. (written against Zig 0.16)
// Build: zig build-exe collatz.zig

const std = @import("std");

const limit = 1000;

fn chainLength(start: u64) u32 {
    var value = start;
    var length: u32 = 1;
    while (value != 1) {
        value = if (value % 2 == 0) value / 2 else value * 3 + 1;
        length += 1;
    }
    return length;
}

pub fn main() void {
    var longest: u64 = 1;
    var best: u32 = 1;

    var start: u64 = 1;
    while (start < limit) : (start += 1) {
        const length = chainLength(start);
        if (length > best) {
            longest = start;
            best = length;
        }
    }

    std.debug.print("{d} {d}\n", .{ longest, best });
}
